ngx_http_sxg_filter_module.c

```diff
@@ -540,6 +574,69 @@ bool is_valid_config(ngx_conf_t* nc, const ngx_http_sxg_srv_conf_t* sc) {
   return valid;
 }
 
+static
+bool append_header(ngx_list_t* headers, const char* key, const char* value) {
+  ngx_table_elt_t* h = ngx_list_push(headers);
+  if (h == NULL) {
+    return false;
+  }
+
+  h->hash = 1;
+  h->key.len = strlen(key);
+  h->key.data = (u_char *) key;
+  h->value.len = strlen(value);
+  h->value.data = (u_char *) value;
+  return true;
+}
+
+static ngx_int_t ngx_http_cert_chain_handler(ngx_http_request_t* req) {
+  // Check the URL is certificate request.
+  ngx_http_sxg_srv_conf_t* ssc =
+      ngx_http_get_module_srv_conf(req, ngx_http_sxg_filter_module);
+  if (ssc->cert_path.len > 0 && req->uri.len == ssc->cert_path.len &&
+      ngx_memcmp(req->uri.data, ssc->cert_path.data, req->uri.len) == 0) {
+    bool refreshed = refresh_if_needed(&ssc->cert_chain);
+    if (refreshed) {
+      ngx_log_error(NGX_LOG_ERR, req->connection->log, 0,
+                    "OCSP Response in Certificate-Chain is refreshed.");
+    }
+    req->headers_out.status = NGX_HTTP_OK;
+    req->headers_out.content_length_n = ssc->cert_chain.serialized_cert_chain.size;
+
+    if (!append_header(&req->headers_out.headers,
+                       "Content-Type", "application/cert-chain+cbor")) {
+      return NGX_HTTP_INTERNAL_SERVER_ERROR;
+    }
+
+    ngx_chain_t* out;
+    if (!make_chain_from_buffer(req, &ssc->cert_chain.serialized_cert_chain, &out)) {
+      ngx_log_error(NGX_LOG_ERR, req->connection->log, 0,
+                    "Failed to generate Cert-Chain.");
+      return NGX_ERROR;
+    }
+    if (ngx_http_next_header_filter(req) != NGX_OK ||
+        ngx_http_next_body_filter(req, out) != NGX_OK) {
+      ngx_log_error(NGX_LOG_ERR, req->connection->log, 0,
+                    "Failed to return payload.");
+      return NGX_ERROR;
+    }
+    return NGX_DONE;
+  }
+  return NGX_OK;
+}
+
+static
+char* ngx_conf_set_cert_chain(ngx_conf_t *cf, ngx_command_t *cmd, void *conf) {
+  ngx_http_sxg_srv_conf_t* ssc = conf;
+  ngx_str_t* args = cf->args->elts;
+  if (ssc->cert_path.data != NULL) {
+    return "is duplicate";
```

Comment:
Should this return a `const char*` since the function returns a string literal?

Reply:
Generally, it should be, but this signature of function must match the `ngx_command_t` requirement which is not `const char*` but `char*`, https://github.com/nginx/nginx/blob/master/src/core/ngx_conf_file.h#L80
In other nginx core code, returning string literal is very usual like [this](https://github.com/phusion/nginx/blob/master/src/http/modules/ngx_http_geoip_module.c#L678)